### PR TITLE
Allow disjoint mutable world access via `EntityMut`

### LINKED ISSUE #4278: Allow users to read / write from all components of a given entity using an `All` `WorldQuery` types
## What problem does this solve or what need does it fill?

From time to time, users want to fetch all of the components from a given set of entities.

To do so, they must use the `EntityRef` or `EntityMut` API, in an exclusive system.

Fundamentally though, there's no reason this must be done in an exclusive system: this does not impact resource data, and we can limit the archetypes affected when combined with query filters.

## What solution would you like?

Create a `All` type that parallels the API of `EntityRef`.

This would enable users to write e.g. `Query<&All, With<Networked>>` in an ordinary system, allowing them to read from any component of the given entity.

This will require some interesting new logic around defining which fetches are needed, but should be relatively straightforward.

## Alternatives considered

Initially, I thought we could use `EntityRef` and `EntityMut` for this by implementing the `WorldQuery` trait on them directly. However, `EntityMut` allows users to add components, which cannot be done in parallel due to the potential impacts on archetypes.

Without the ability to reuse the APIs, IMO it makes more sense just to create a new type and use the standard & vs &mut distinction.

Alternatively alternatively, we could instead use one of the more 🧪 ideas to allow for instant component addition / removals in parallel systems and then just use EntityRef and EntityMut, likely by deferring the archetype updates until the next sync point and storing new component data in a scratch space of sorts.

### LINKED ISSUE #5504: Iterate ALL Entities in world (get EntityRef/EntityMut)
## What problem does this solve or what need does it fill?

There are cases where direct access to EntityRef/EntityMut is needed in an iterating context

## What solution would you like?

An archetype-based solution, ideally we could have a &EntityRef or &mut EntityMut in queries, which would automatically imply exclusive world access thinking. (https://github.com/bevyengine/bevy/discussions/5496#discussioncomment-3288227)

## What alternative(s) have you considered?

Iterating Entity (the ID) and calling World::entity. This solution works, but it is unintuitive and suffers from worse performance than a built-in solution for iterating EntityRef/EntityMut

## Additional context

Original discussion: https://github.com/bevyengine/bevy/discussions/5496

### LINKED ISSUE #9422: `EntityRef as WorldQuery` is unsound.
## Bevy version
 Main branch.

## What you did

```rust
    #[test]
    fn unsound_disjoint_mutable_access() {
        fn unsound_system(
            mut query: Query<(Entity, &mut TestComponent)>,
            refs: Query<super::EntityRef, Without<TestComponent>>,
        ) {
            let (id, mut c) = query.iter_mut().next().unwrap();
            let world = refs.iter().next().unwrap().world();

            // We now have mutable and shared access to the same component at the same time.
            let c2 = world.entity(id).get::<TestComponent>().unwrap();
            *c = TestComponent(1);
            println!("{c2:?}");
        }

        let mut world = World::new();
        world.spawn(TestComponent(0));
        world.spawn_empty();

        let mut schedule = Schedule::new();
        schedule.add_systems(unsound_system);

        schedule.run(&mut world);
    }
```

## What went wrong

This should panic at runtime since the accesses of the two system params conflict. Instead, it runs and allows aliased mutable access.

### PR DESCRIPTION
**Pull request by JoJoJet**
# Objective

Fix #4278
Fix #5504
Fix #9422

Provide safe ways to borrow an entire entity, while allowing disjoint mutable access.  `EntityRef` and `EntityMut` are not suitable for this, since they provide access to the entire world -- they are just helper types for working with `&World`/`&mut World`.

This has potential uses for reflection and serialization

## Solution

Remove `EntityRef::world`, which allows it to soundly be used within queries.

`EntityMut` no longer supports structural world mutations, which allows multiple instances of it to exist for different entities at once. Structural world mutations are performed using the new type `EntityWorldMut`.

```rust
fn disjoint_system(
     q2: Query<&mut A>,
     q1: Query<EntityMut, Without<A>>,
) { ... }

let [entity1, entity2] = world.many_entities_mut([id1, id2]);
*entity1.get_mut::<T>().unwrap() = *entity2.get().unwrap();

for entity in world.iter_entities_mut() {
    ...
}
```

---

## Changelog

- Removed `EntityRef::world`, to fix a soundness issue with queries.
+ Removed the ability to structurally mutate the world using `EntityMut`, which allows it to be used in queries.
+ Added `EntityWorldMut`, which is used to perform structural mutations that are no longer allowed using `EntityMut`.

## Migration Guide

**Note for maintainers: ensure that the guide for #9604 is updated accordingly.**

Removed the method `EntityRef::world`, to fix a soundness issue with queries. If you need access to `&World` while using an `EntityRef`, consider passing the world as a separate parameter.

`EntityMut` can no longer perform 'structural' world mutations, such as adding or removing components, or despawning the entity. Additionally, `EntityMut::world`, `EntityMut::world_mut` , and `EntityMut::world_scope` have been removed.
Instead, use the newly-added type `EntityWorldMut`, which is a helper type for working with `&mut World`.

### CODE REVIEW

File: crates/bevy_ecs/src/query/access.rs [hunk 1]
```diff
@@ -113,39 +116,58 @@ impl<T: SparseSetIndex> Access<T> {
...
+        self.writes_all = true;
+    }
+
     /// Returns `true` if this has access to all indexed elements (i.e. `&World`).
     pub fn has_read_all(&self) -> bool {
-        self.reads_all
+        self.reads_all || self.writes_all
+    }
+
+    /// Returns `true` if this has write access to all indexec elements (i.e `EntityBorrowMut`).
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:154]:
```suggestion
    /// Returns `true` if this has write access to all indexed elements (i.e `EntityBorrowMut`).
```

**Review by alice-i-cecile — COMMENTED** [review]
Part of me feels that `EntityRef` and `EntityMut` should be removed completely in this PR. This feels like a much more helpful abstraction that serves a very similar role.

Why is that a bad idea?

File: crates/bevy_ecs/src/world/entity_borrow.rs [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+    #[test]
+    fn borrow_mut_compatible() {
+        fn borrow_mut_system(_: Query<(Entity, EntityBorrowMut)>) {}
+
+        assert_is_system(borrow_mut_system);
+    }
+
+    #[test]
+    #[should_panic]
+    fn borrow_mut_incompatible1() {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/entity_borrow.rs:418]: These tests should have better names, even if they're more verbose. Maybe `borrow_mut_incompatible_with_read_only_component` and so on?

File: crates/bevy_ecs/src/world/mod.rs [hunk 3]
```diff
@@ -385,12 +508,38 @@ impl World {
...
                     };
 
-                    // SAFETY: entity exists and location accurately specifies the archetype where the entity is stored
+                    // SAFETY: entity exists and location accurately specifies the archetype where the entity is stored.
                     unsafe { EntityRef::new(self, entity, location) }
                 })
         })
     }
 
+    /// Returns a mutable iterator over current entities.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/mod.rs:517]:
```suggestion
    /// Returns a mutable iterator over all entities in the `World`.
```

File: crates/bevy_ecs/src/world/mod.rs [hunk 4]
```diff
@@ -385,12 +508,38 @@ impl World {
...
+    pub fn iter_entities_mut(&mut self) -> impl Iterator<Item = EntityBorrowMut<'_>> + '_ {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/mod.rs:518]: Should we have an immutable equivalent?

**Comment by alice-i-cecile** [comment]
Generally impressed: this is an extremely useful bit of API (especially for reflection-adjacent tasks) and is well-made. The question about `EntityRef` and `EntityMut` is my main concern: they seem rather duplicative and less elegant.

If we want to keep them, I'd be interested in a way to go *back* to the `World`-containing variants, by passing in `&World` or `&mut World` respectively.

**Comment by JoJoJet** [comment]
> Part of me feels that EntityRef and EntityMut should be removed completely in this PR. This feels like a much more helpful abstraction that serves a very similar role.
> Why is that a bad idea?

`EntityMut` is still useful on its own. It allows you to insert and remove components, which `EntityBorrowMut` cannot do.

I could see an argument for removing `EntityRef` though. I mostly left it because `EntityRef::world` is mildly useful sometimes.

File: crates/bevy_ecs/src/world/mod.rs (hunk 4 above)
Comment (JoJoJet) [reply crates/bevy_ecs/src/world/mod.rs:518]: There already is one :)

**Comment by JoJoJet** [comment]
I've updated this to remove the `WorldQuery` impl for `EntityRef`, in order to fix #9422.

**Review by alice-i-cecile — APPROVED** [review]
Right: `EntityMut` is genuinely useful. Alright, let's move forward like this for now.

File: crates/bevy_ecs/src/world/entity_borrow.rs [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+    pub fn get_mut_by_id(&mut self, component_id: ComponentId) -> Option<MutUntyped<'_>> {
+        // SAFETY:
+        // - `&mut self` ensures that no references exist to this entity's components.
+        // - `as_unsafe_world_cell` gives mutable permission for all components on this entity
+        unsafe { self.0.get_mut_by_id(component_id) }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (iiYese) [inline crates/bevy_ecs/src/world/entity_borrow.rs:377]: Shouldn't there also be a panicing test for `Query<EntityBorrow>` + `Query<&mut A>`?

**Comment by iiYese** [comment]
> I mostly left it because `EntityRef::world` is mildly useful sometimes

I would like identified cases of when. The API duplication is getting ridiculous now
- `&mut World`
- `EntityRef`
- `EntityMut`
- `EntityBorrow`
- `EntityBorrowMut`

Should try not to bloat this further until refactors on the horizon are possible. If no good argument can be made for or against keeping `EntityRef` then a push for its removal should be made to find one.

**Comment by JoJoJet** [comment]
> > I mostly left it because `EntityRef::world` is mildly useful sometimes
> 
> I would like identified cases of when.

I like to think of `EntityRef` as a performance-optimized `(Entity, &World)` tuple, which can be useful any time you're working with direct world access (I have used it for this before). If we removed `EntityRef::world`, then a bunch of users would have to change some method signatures from `fn(EntityRef)` to `fn(EntityRef, &World)` -- not the biggest loss but still annoying.

That said, I'm open to doing it if we decide it's a good idea. I think the path with the least API breakages would be to remove `EntityRef`, and rename `EntityBorrow` ->  `EntityRef`. This would open the question of what to call `EntityBorrowMut`, which would seem out of place without a corresponding `EntityBorrow`.

**Comment by JoJoJet** [comment]
I've added some unit tests that are known to fail due to limitations in our current access model. Since the issue is with overly restrictive accesses, it will not cause undefined behavior, so I think it is okay to fix this issue later. Also, note that this issue already existed on the main branch with `EntityRef` queries.

**Comment by JoJoJet** [comment]
I have merged `EntityBorrow` into `EntityRef`, and removed `EntityRef::world`.

**Comment by JoJoJet** [comment]
I have also renamed `EntityMut` to `EntityWorldMut`, and renamed `EntityBorrowMut` to `EntityMut`. Since I've made so many changes, I've updated the PR description to give an overview of the current state of this PR.

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 6]
```diff
@@ -166,35 +149,263 @@ impl<'w> EntityRef<'w> {
...
+/// # Examples
+///
+/// Disjoint mutable access.
+///
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// # #[derive(Component)] pub struct A;
+/// fn disjoint_system(
+///     query1: Query<EntityRef, With<A>>,
+///     query2: Query<EntityRef, Without<A>>,
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/world/entity_ref.rs:206]:
```suggestion
///     query1: Query<EntityMut, With<A>>,
///     query2: Query<EntityMut, Without<A>>,
```

File: crates/bevy_ecs/src/world/entity_ref.rs [hunk 7]
```diff
@@ -1369,4 +1576,113 @@ mod tests {
...
+
+    #[derive(Component)]
+    struct A;
+
+    #[derive(Resource)]
+    struct R;
+
+    #[test]
+    fn disjoint_access() {
+        fn disjoint_readonly(_: Query<EntityRef, With<A>>, _: Query<EntityRef, Without<A>>) {}
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/world/entity_ref.rs:1588]:
```suggestion
        fn disjoint_readonly(_: Query<EntityMut, With<A>>, _: Query<EntityRef, Without<A>>) {}
```

**Review by maniwani — APPROVED** [review]
This looks good to me. I mostly wanted to make sure the access is sound. 

I think renaming the existing types is a little strange, but ig it's better for queries to get the types with shorter names. It's fine if there's some type duplication between "queried" and "direct-world" access though. We may find a need for an `EntityWorldRef` later.

**Comment by alice-i-cecile** [comment]
I think there are subtle merge conflicts here @JoJoJet

**Pull request merged**